### PR TITLE
Improve the performance of movie2d.py script

### PR DESCRIPTION
**Pull request by tbhaxor**
## PR Summary

This PR solves a lot of things in the movie2d. Lets go one by one

1. Use a parallel processing using ProcessPoolExecutor. I tested with 10 workers, and it working fine. Although, I am printing the matplotlib threading warning. This improved the performance as shown below
    | Old Code | New Code |
    | :----:   | :----:   |
    |  |  | 

   > **Note** Right click on the images and open them in new-tab to get more clear picture
2. Optionally add time-step to show as the axes title
 
    

3. Check if field exists in the data for each file and break the for-loop on error
4. Remove unused and unaffecting (_idk if that is even a work_) functions
5. Accept the `-d` / `--output-directory` parameter to save the files in the specified directory. It is using pathlib.Path so the functions are pretty simplified as well.
6. Provide switching of `--debug` option from the arguments
7. Generate gif or mp4 movie from the output file images

Motivation: My simulation code used to run faster than movie generation process.

## PR Checklist

- [x] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Comment by tbhaxor** [comment]
CPP linter didnt report any error in the files and also all the formatting related to `scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py` file is fixed in 8e273e24234e3b9ca1aa3426649be577f0783995

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:44]: You may have to change this back---black will want a particular formatting.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: This could be made more performant by having the worker in each pool also load the dump file and the dataset. A wrapper function around `plot_dump` and `phdf ` would need to be written. You could also use `pool.map` instead of `pool.submit`, which would be a little bit more concise.

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Thanks for submitting this! While having the ability to generate frames in parallel with processor pool is a nice addition, I do not think it should be the same script--as  people often use this script as a one-off to generate a few images or to just generate a single frame. My suggestion: rename `movie2d.py` to `snap2d.py`, and then add a movie script that imports `make_frame` and generates frames in parallel as you do here. This has a few advantages:
- The extra complexity of parallel processing is contained, so that folks can still read and use this plot without having to think about processor pools.
- There is a separation of concerns, which makes for more clean and maintainable code.
- In that model you could also call ffmpeg to generate the movie at the end in a seamless way.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Why use this? It would probably be better to simply pull the time of each frame out of the dump file.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Does the `/` operator append paths? I didn't know that.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Since the only change here is that it runs in serial, I would call this `--serial` not `--debug` unless it makes additional changes.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:75]: I like the ability to output to a separate directory! But while we're at it, a filename prefix would probably be just as if not more useful... e.g., `--prefix run_name` would produce files like `run_name_0000.png`.

**Review by Yurlungur — COMMENTED** [review]
A few more nitpicks below.

It's also worth noting that this script doesn't need to be super feature-complete. It's mostly a utility we use for small things.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: I think this flag actually sets it to output the exploded plots that show ghost zones (which I called debug plots since they are mainly used to debug issues with downstream codes/Parthenon, potentially confusingly). It doesn't look like it actually gets run on a single thread when this argument is set (see line 222). 

I also think `args.debug` probably shouldn't set the level of logging at line 183.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Yes it does that. Also it is well documented here https://docs.python.org/3/library/pathlib.html#operators

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:44]: I tried running this but black didn't complain or changed this. Either way it looks more readable to me.

Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:44]: Do you still want me to revert to the older version for tc and vc?

**Comment by tbhaxor** [comment]
> It's also worth noting that this script doesn't need to be super feature-complete. It's mostly a utility we use for small things.

Even I dont wanted to work on it earlier, but the time it takes to generate output for images is not efficient.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Yeah I didn't know about it. Thanks for the information. It is fixed now

Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Yeah I didn't know about it. Thanks for the information. It is fixed now

**Comment by tbhaxor** [comment]
> It's also worth noting that this script doesn't need to be super feature-complete. It's mostly a utility we use for small things.

Even I dont wanted to work on it earlier, but the time it takes to generate output for images is not efficient.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: > I also think args.debug probably shouldn't set the level of logging at line 183.

Yes makes sense to me. It is fixed now

Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: > It doesn't look like it actually gets run on a single thread when this argument is set (see line 222).

That was a typo. I have also fixed that

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:75]: Added `--prefix`. By default it is `default_run`. You can test it

**Comment by tbhaxor** [comment]
@Yurlungur Under which title do you think I should document this change in the CHANGELOG.md

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: @Yurlungur I couldn't get what you are saying. Could you give more information here?

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:44]: While running the black, I couldnt get any warning/error or reformat updates from the black. It is more readable to me now. 

Do you want me to change it back to previous version?

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Ohk, but how can we check for the field and break the loop

```py
if args.field not in data.Variables:
                logging.error(
                    f'No such field "{args.field}" in {file_name}. \
This will lead to stop further processing'
                )
                break
```

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:44]: I prefer the older formatting but if black is happy, it's fine.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Your version produces a cleaner error message, so that's fine. It's just a little bit more concise to construct an array of args to pass in and then `pool.map(make_frame_worker, args)`. You'll still get a useful error message out, but it'll be produced asynchronously by the worker that failed.

Although actually I just thought of something: Currently the `print(data)` line means `movie2d` outputs all the available variables so if I for example call
```python
movie2d.py nonexistent_var my_run.out0.00000.phdf
```
it will output the variables available to plot before the code dies. Your version does not do this. Any error message that runs the movie2d script should provide this so that users can easily get some introspection about the mesh.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: :+1:

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: :+1:

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:75]: I would suggest a default of the empty string for the `prefix`.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: :+1: Probably the help message should not say anything about serial vs parallel then.

**Comment by Yurlungur** [comment]
> Even I dont wanted to work on it earlier, but the time it takes to generate output for images is not efficient.

Has the movie script impeded your workflow? I can run through several hundred outputs in a pretty short amount of time, even in serial.

> @Yurlungur Under which title do you think I should document this change in the CHANGELOG.md

`Changed` I think

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: ACK

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Yes, fixed already

**Comment by tbhaxor** [comment]
Now it also generates movies 

Screenshot_20230405_180336

You can disable it using `--no-render` option

**Comment by tbhaxor** [comment]
> Has the movie script impeded your workflow?

Not that much but on 1000+ files it is very slow

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: I am now showing all the variables

Screenshot_20230405_180900

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:79]: :+1:

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: I would switch the logic here. I.e., `--render` should be opt in and `--no-render` should be the default. I often use the script to generate a single frame, and I'm sure I'm not the only one.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Regex and auxiliary variable here seems unnecessarily fancy. Why not just do the following?
```suggestion
            name = "{}_{:04d}.png".format(args.prefix, dump_id).strip()
```
sometimes old-style string-formatting is more clear and concise than f-strings. (Also the underscore for the prefix is probably desirable.)

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: `dump_id` is a little confusing as the dumps may not align to the frame numbers. I'd say `frame_index` or `frame_idx` or something like that.

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Thanks for all these changes. Just a few more nitpicks below. It sounds like you prefer to do everything in the same script, rather than split it into a ` snap` script and a `movie` script?

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:287]: Sometimes systems have `libav` and not `ffmpeg`. So it might be worth adding an option to switch ``ffmpeg` for `libav`.

Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:287]: I have never heard of libav, but will explore an use it. Thanks for suggesting @Yurlungur

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Roger that 👍🏻

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py:287]: :+1:

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: It may be better to have a more descriptive help message here, i.e. `help="Make plots with an exploded grid and including ghost zones. (default: false)"`. Maybe the option should be `--debug-plot` or something, just `--debug` is a bit confusing.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: I agree that `--no-render` should be the default.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]:
```suggestion
```

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]:
```suggestion
                    f'No such field "{args.field}" in {file_name}. Further processing stopped.'
```

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Maybe I am missing something, but it looks like `components` is not getting passed here. I think that was maybe true of the old code as well, but I think it was a bug.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (lroberts36) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]:
```suggestion
        logger.info("All frames produced.")
```

**Review by lroberts36 — APPROVED** [review]
This LGTM, but I agree with @Yurlungur that not producing a movie should be the default.

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Fixed

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Done

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: It is not even passed in the develop branch. This is the latest commit in the develop branch.

https://github.com/parthenon-hpc-lab/parthenon/blob/a61999403a09a2946bd6ae40b5140169833b844a/scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py#L185

Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: I think what @lroberts36 is saying is that this is a bug in `develop`

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Done, change to `--render`

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: I see, should I add it then?

Comment (Yurlungur) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: assume it doesn't break functionality, yes please.

Comment (lroberts36) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: yes, I think so

Comment (tbhaxor) [reply scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: @Yurlungur

Done, sir

File: scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/movie2d.py]: Sorry to change my mind on you. Probably you were right the first time, no need for the underscore.
```suggestion
            name = "{}{:04d}.png".format(args.prefix, frame_id).strip()
```

**Review by Yurlungur — APPROVED** [review]
Ok I think this is good to go, pending one last suggestion below. the libav vs ffmpeg thing can wait. Thanks for bearing with all the pickyness on this one.

**Comment by Yurlungur** [comment]
Ok I'd like to make sure the script still works on some edge cases, which I will play with tomorrow (hopefully). After that, since @lroberts36 and I both approve, I can merge it.

**Comment by tbhaxor** [comment]
> Ok I think this is good to go, pending one last suggestion below. the libav vs ffmpeg thing can wait. Thanks for bearing with all the pickyness on this one.

No problem at all sir 😄

**Comment by tbhaxor** [comment]
@Yurlungur any updates?

**Comment by Yurlungur** [comment]
Ok I'm happy. I'm enabling automerge.

**Comment by tbhaxor** [comment]
There are some errors in the cuda11.6 container setup here thats causing auto merge to fail.

**Comment by Yurlungur** [comment]
The CI machine currently is having issues. Should be resolved soon.

**Comment by Yurlungur** [comment]
@tbhaxor I forgot our C++ and python formatting pipeline doesn't work on forks. Can you confirm you ran black on the script?

**Comment by tbhaxor** [comment]
> @tbhaxor I forgot our C++ and python formatting pipeline doesn't work on forks. Can you confirm you ran black on the script?

Reformatted 181d896

**Comment by Yurlungur** [comment]
Great thanks. I will force merge since all the other tests passed.

**Pull request merged**
